node/src/protocol/presignature.rs

```diff
@@ -0,0 +1,239 @@
+        threshold: usize,
+        triple0: Triple,
+        triple1: Triple,
+        public_key: &PublicKey,
+        private_share: &PrivateKeyShare,
+    ) -> Result<PresignatureGenerator, InitializationError> {
+        let protocol = Box::new(cait_sith::presign(
+            participants,
+            me,
+            PresignArguments {
+                triple0: (triple0.share, triple0.public),
+                triple1: (triple1.share, triple1.public),
+                keygen_out: KeygenOutput {
+                    private_share: *private_share,
+                    public_key: *public_key,
+                },
+                threshold,
+            },
+        )?);
+        Ok(PresignatureGenerator {
+            protocol,
+            triple0: triple0.id,
+            triple1: triple1.id,
+        })
+    }
+
+    /// Starts a new presignature generation protocol.
+    pub fn generate(
+        &mut self,
+        triple0: Triple,
+        triple1: Triple,
+        public_key: &PublicKey,
+        private_share: &PrivateKeyShare,
+    ) -> Result<(), InitializationError> {
+        let id = rand::random();
+        tracing::info!(id, "starting protocol to generate a new presignature");
+        let generator = Self::generate_internal(
+            &self.participants,
+            self.me,
+            self.threshold,
+            triple0,
+            triple1,
+            public_key,
+            private_share,
+        )?;
+        self.generators.insert(id, generator);
+        Ok(())
+    }
+
+    /// Ensures that the presignature with the given id is either:
+    /// 1) Already generated in which case returns `None`, or
+    /// 2) Is currently being generated by `protocol` in which case returns `Some(protocol)`, or
+    /// 3) Has never been seen by the manager in which case start a new protocol and returns `Some(protocol)`, or
+    /// 4) Depends on triples (`triple0`/`triple1`) that are unknown to the node
+    // TODO: What if the presignature completed generation and is already spent?
+    pub fn get_or_generate(
+        &mut self,
+        id: PresignatureId,
+        triple0: TripleId,
+        triple1: TripleId,
+        triple_manager: &mut TripleManager,
+        public_key: &PublicKey,
+        private_share: &PrivateKeyShare,
+    ) -> Result<Option<&mut PresignatureProtocol>, InitializationError> {
+        if self.presignatures.contains_key(&id) {
+            Ok(None)
+        } else {
+            match self.generators.entry(id) {
+                Entry::Vacant(entry) => {
+                    tracing::info!(id, "joining protocol to generate a new presignature");
+                    let triple0 = match triple_manager.take(triple0) {
+                        Some(triple0) => triple0,
+                        None => {
+                            tracing::warn!(triple_id = triple0, "triple0 is missing, can't join");
+                            return Ok(None);
+                        }
+                    };
```

Comment:
nit: let-else could be used here
```suggestion
                    let Some(triple0) = triple_manager.take(triple0) else {
                        tracing::warn!(triple_id = triple0, "triple0 is missing, can't join");
                        return Ok(None);
                    };
```